File: Source/OC/UIView-WZB/UIView+WZB.m
//
//  UIView+WZB.m
//  WZBListView
//
//  Created by 王振标 on 16/6/13.
//  Copyright © 2016年 王振标. All rights reserved.
//

#import "UIView+WZB.h"

#define SINGLE_LINE_WIDTH           (1 / [UIScreen mainScreen].scale)
#define SINGLE_LINE_ADJUST_OFFSET   ((1 / [UIScreen mainScreen].scale) / 2)

#define WZBTag 155158

@implementation UIView (WZB)

/**
 * 创建一个表格
 * columns：列数
 * rows：行数
 * data：数据
 */
- (void)wzb_drawListWithRect:(CGRect)rect columns:(NSInteger)columns rows:(NSInteger)rows datas:(NSArray *)datas
{
    [self wzb_drawListWithRect:rect columns:columns rows:rows datas:datas colorInfo:nil];
}

/**
 * 创建一个表格
 * columns：列数
 * rows：行数
 * data：数据
 * lineInfo：行信息，传入格式：@{@"0" : @"3"}意味着第一行创建3个格子
 */
- (void)wzb_drawListWithRect:(CGRect)rect columns:(NSInteger)columns rows:(NSInteger)rows datas:(NSArray *)datas columnsInfo:(NSDictionary *)columnsInfo
{
    [self wzb_drawListWithRect:rect columns:columns rows:rows datas:datas colorInfo:nil columnsInfo:columnsInfo];
}

/**
 * 创建一个表格
 * line：列数
 * rows：行数
 * data：数据
 * colorInfo：颜色信息，传入格式：@{@"0" : [UIColor redColor]}意味着第一个格子文字将会变成红色
 */
- (void)wzb_drawListWithRect:(CGRect)rect columns:(NSInteger)columns rows:(NSInteger)rows datas:(NSArray *)datas colorInfo:(NSDictionary *)colorInfo
{
    [self wzb_drawListWithRect:rect columns:columns rows:rows datas:datas colorInfo:colorInfo columnsInfo:nil];
}

/**
 * 创建一个表格
 * columns：列数
 * rows：行数
 * data：数据
 * colorInfo：颜色信息，传入格式：@{@"0" : [UIColor redColor]}意味着第一个格子文字将会变成红色
 * columnsInfo：行信息，传入格式：@{@"0" : @"3"}意味着第一行创建3个格子
 */
- (void)wzb_drawListWithRect:(CGRect)rect columns:(NSInteger)columns rows:(NSInteger)rows datas:(NSArray *)datas colorInfo:(NSDictionary *)colorInfo columnsInfo:(NSDictionary *)columnsInfo
{
    [self wzb_drawListWithRect:rect columns:columns rows:rows datas:datas colorInfo:colorInfo columnsInfo:columnsInfo backgroundColorInfo:nil];
}

/**
 * 创建一个表格
 * columns：列数
 * rows：行数
 * data：数据
 * colorInfo：颜色信息，传入格式：@{@"0" : [UIColor redColor]}意味着第一个格子文字将会变成红色
 * columnsInfo：行信息，传入格式：@{@"0" : @"3"}意味着第一行创建3个格子
 * backgroundColorInfo：行信息，传入格式：@{@"0" : [UIColor redColor]}意味着第一个格子背景颜色变成红色
 */
- (void)wzb_drawListWithRect:(CGRect)rect columns:(NSInteger)columns rows:(NSInteger)rows datas:(NSArray *)datas colorInfo:(NSDictionary *)colorInfo columnsInfo:(NSDictionary *)columnsInfo backgroundColorInfo:(NSDictionary *)backgroundColorInfo
{
    NSInteger index = 0;
    CGFloat x = rect.origin.x;
    CGFloat y = rect.origin.y;
    CGFloat h = (1.0) * rect.size.height / rows;
    NSInteger newLine = 0;
    for (NSInteger i = 0; i < rows; i++) {
        
        // 判断合并单元格
        if (columnsInfo) {
            for (NSInteger a = 0; a < columnsInfo.allKeys.count; a++) {
                
                // 新的列数
                NSInteger newColumn = [columnsInfo.allKeys[a] integerValue];
                if (i == newColumn) {
                    newLine = [columnsInfo[columnsInfo.allKeys[a]] integerValue];
                } else {
                    newLine = columns;
                }
            }
        } else {
            newLine = columns;
        }
        
        
        for (NSInteger j = 0; j < newLine; j++) {
            
            // 线宽
            CGFloat w = (1.0) * rect.size.width / newLine;
            CGRect frame = (CGRect){x + w * j, y + h * i, w, h};
            
            // 画线
            [self wzb_drawRectWithRect:frame];
            
            // 创建label
            UILabel *label = [[UILabel alloc] initWithFrame:frame];
            
            // 文字居中
            label.textAlignment = NSTextAlignmentCenter;
            [self addSubview:label];
            
            // 判断文字颜色
            UIColor *textColor = [colorInfo objectForKey:[NSString stringWithFormat:@"%zd", index]];
            if (!textColor) {
                textColor = [UIColor grayColor];
            }
            label.textColor = textColor;
            
            // 判断背景颜色
            UIColor *backgroundColor = [backgroundColorInfo objectForKey:[NSString stringWithFormat:@"%zd", index]];
            if (!backgroundColor) {
                backgroundColor = [UIColor clearColor];
            }
            label.backgroundColor = backgroundColor;
            
            // 字体大小
            label.font = [UIFont systemFontOfSize:13];
            
            // label文字
            label.text = datas[index];
            
            // label的tag值
            label.tag = WZBTag + index;
            index++;
        }
    }
}

- (void)wzb_drawRectWithRect:(CGRect)rect {
    CGFloat x = rect.origin.x;
    CGFloat y = rect.origin.y;
    CGFloat w = (1.0) * rect.size.width;
    CGFloat h = (1.0) * rect.size.height;
    
    if (((int)(y * [UIScreen mainScreen].scale) + 1) % 2 == 0) {
        y += SINGLE_LINE_ADJUST_OFFSET;
    }
    if (((int)(x * [UIScreen mainScreen].scale) + 1) % 2 == 0) {
        x += SINGLE_LINE_ADJUST_OFFSET;
    }
    
    [self wzb_drawLineWithFrame:(CGRect){x, y, w, 1} type:1];
    [self wzb_drawLineWithFrame:(CGRect){x + w, y, 1, h} type:2];
    [self wzb_drawLineWithFrame:(CGRect){x, y + h, w, 1} type:1];
    [self wzb_drawLineWithFrame:(CGRect){x, y, 1, h} type:2];
}

- (void)wzb_drawLineWithFrame:(CGRect)frame lineType:(WZBLineType)lineType color:(UIColor *)color lineWidth:(CGFloat)lineWidth {
    
    // 创建贝塞尔曲线
    UIBezierPath *linePath = [[UIBezierPath alloc] init];
    
    // 线宽
    linePath.lineWidth = lineWidth;
    
    // 起点
    [linePath moveToPoint:CGPointMake(0, 0)];
    
    // 重点：判断是水平方向还是垂直方向
    [linePath addLineToPoint: lineType == WZBLineHorizontal ? CGPointMake(frame.size.width, 0) : CGPointMake(0, frame.size.height)];
    
    // 创建CAShapeLayer
    CAShapeLayer *lineLayer = [CAShapeLayer layer];
    
    // 颜色
    lineLayer.strokeColor = color.CGColor;
    // 宽度
    lineLayer.lineWidth = lineWidth;
    
    // frame
    lineLayer.frame = frame;
    
    // 路径
    lineLayer.path = linePath.CGPath;
    
    // 添加到layer上
    [self.layer addSublayer:lineLayer];
}

- (void)wzb_drawLineWithFrame:(CGRect)frame type:(NSInteger)type color:(UIColor *)color {
    [self wzb_drawLineWithFrame:frame lineType:type color:color lineWidth:0.7];
}

- (void)wzb_drawLineWithFrame:(CGRect)frame type:(NSInteger)type {
    [self wzb_drawLineWithFrame:frame type:type color:[UIColor blackColor]];
}

// 根据tag拿到对应的label
- (UILabel *)getLabelWithIndex:(NSInteger)index {
    
    // 为了防止self的subviews又重复的tag，拿到第一个
    for (UIView *v in self.subviews) {
        
        // 判断是否为label类
        if ([v isKindOfClass:[UILabel class]]) {
            
            // 强制转换
            UILabel *label = (UILabel *)v;
            if (v.tag == index + WZBTag) {
                return label;
            }
        }
    }
    return [self viewWithTag:index + WZBTag];
}

@end

File: Source/Swift/UIView+WZB/UIView+WZB.swift
import ObjectiveC
import UIKit

private func wzbDefaultTextColor() -> UIColor {
    if #available(iOS 13.0, *) {
        return .secondaryLabel
    }
    return .gray
}

private func wzbDefaultSeparatorColor() -> UIColor {
    if #available(iOS 13.0, *) {
        return .separator
    }
    return UIColor(white: 0.82, alpha: 1)
}

@objc public enum WZBLineType: Int {
    case vertical
    case horizontal
}

private final class WZBGridStorage: NSObject {
    var labels: [UILabel] = []
    var lineLayers: [CAShapeLayer] = []
}

private final class WZBScrollableGridStorage: NSObject {
    weak var containerView: UIView?
}

private enum WZBAssociatedKeys {
    static var gridStorage: UInt8 = 0
    static var scrollableGridStorage: UInt8 = 0
}

private extension UIView {
    var wzb_gridStorage: WZBGridStorage {
        if let storage = objc_getAssociatedObject(self, &WZBAssociatedKeys.gridStorage) as? WZBGridStorage {
            return storage
        }

        let storage = WZBGridStorage()
        objc_setAssociatedObject(
            self,
            &WZBAssociatedKeys.gridStorage,
            storage,
            .OBJC_ASSOCIATION_RETAIN_NONATOMIC
        )
        return storage
    }

    var wzb_hairlineWidth: CGFloat {
        1 / UIScreen.main.scale
    }

    func wzb_clearGridArtifacts() {
        wzb_gridStorage.labels.forEach { $0.removeFromSuperview() }
        wzb_gridStorage.lineLayers.forEach { $0.removeFromSuperlayer() }
        wzb_gridStorage.labels.removeAll()
        wzb_gridStorage.lineLayers.removeAll()
    }

    func wzb_resolvedColumns(
        defaultColumns: Int,
        row: Int,
        columnsInfo: [Int: Int]?
    ) -> Int {
        max(columnsInfo?[row] ?? defaultColumns, 1)
    }

    func wzb_resolvedRowHeights(rows: Int, rowHeights: [CGFloat]?, totalHeight: CGFloat) -> [CGFloat] {
        guard let rowHeights, rowHeights.count == rows else {
            return Array(repeating: totalHeight / CGFloat(rows), count: rows)
        }

        let sanitizedHeights = rowHeights.map { max($0, 1) }
        let total = sanitizedHeights.reduce(0, +)
        guard total > 0 else {
            return Array(repeating: totalHeight / CGFloat(rows), count: rows)
        }

        return sanitizedHeights.map { totalHeight * ($0 / total) }
    }

    func wzb_resolvedColumnWidths(
        defaultColumns: Int,
        row: Int,
        columnsInfo: [Int: Int]?,
        columnWidthInfo: [Int: [CGFloat]]?,
        totalWidth: CGFloat
    ) -> [CGFloat] {
        let currentColumns = wzb_resolvedColumns(defaultColumns: defaultColumns, row: row, columnsInfo: columnsInfo)

        guard
            let widthWeights = columnWidthInfo?[row],
            widthWeights.count == currentColumns
        else {
            return Array(repeating: totalWidth / CGFloat(currentColumns), count: currentColumns)
        }

        let sanitizedWeights = widthWeights.map { max($0, 1) }
        let total = sanitizedWeights.reduce(0, +)
        guard total > 0 else {
            return Array(repeating: totalWidth / CGFloat(currentColumns), count: currentColumns)
        }

        return sanitizedWeights.map { totalWidth * ($0 / total) }
    }

    func wzb_makeLabel(frame: CGRect, text: Any?, textColor: UIColor, backgroundColor: UIColor) -> UILabel {
        let label = UILabel(frame: frame.integral)
        label.textAlignment = .center
        label.numberOfLines = 0
        label.adjustsFontSizeToFitWidth = true
        label.minimumScaleFactor = 0.7
        label.font = .systemFont(ofSize: 13)
        label.textColor = textColor
        label.backgroundColor = backgroundColor
        label.text = text.map { String(describing: $0) } ?? ""
        return label
    }

    func wzb_storeLabel(_ label: UILabel) {
        wzb_gridStorage.labels.append(label)
        addSubview(label)
    }

    func wzb_storeLineLayer(_ layer: CAShapeLayer) {
        wzb_gridStorage.lineLayers.append(layer)
        self.layer.addSublayer(layer)
    }

    func wzb_drawList(
        with rect: CGRect,
        defaultColumns: Int,
        rows: Int,
        datas: [Any],
        colorInfo: [Int: UIColor]?,
        columnsInfo: [Int: Int]?,
        columnWidthInfo: [Int: [CGFloat]]?,
        rowHeights: [CGFloat]?,
        backgroundColorInfo: [Int: UIColor]?
    ) {
        guard defaultColumns > 0, rows > 0, rect.width > 0, rect.height > 0 else { return }

        wzb_clearGridArtifacts()

        let resolvedRowHeights = wzb_resolvedRowHeights(rows: rows, rowHeights: rowHeights, totalHeight: rect.height)
        var dataIndex = 0
        var currentY = rect.minY

        for row in 0 ..< rows {
            let currentRowHeight = resolvedRowHeights[row]
            let columnWidths = wzb_resolvedColumnWidths(
                defaultColumns: defaultColumns,
                row: row,
                columnsInfo: columnsInfo,
                columnWidthInfo: columnWidthInfo,
                totalWidth: rect.width
            )

            var currentX = rect.minX
            for columnWidth in columnWidths {
                let cellFrame = CGRect(
                    x: currentX,
                    y: currentY,
                    width: columnWidth,
                    height: currentRowHeight
                )

                wzb_drawGridBorder(for: cellFrame)

                let label = wzb_makeLabel(
                    frame: cellFrame,
                    text: dataIndex < datas.count ? datas[dataIndex] : nil,
                    textColor: colorInfo?[dataIndex] ?? wzbDefaultTextColor(),
                    backgroundColor: backgroundColorInfo?[dataIndex] ?? .clear
                )

                wzb_storeLabel(label)
                dataIndex += 1
                currentX += columnWidth
            }

            currentY += currentRowHeight
        }
    }
}

public extension UIView {
    @objc(wzb_drawListWithRect:columns:rows:datas:)
    func wzb_drawList(with rect: CGRect, columns: Int, rows: Int, datas: [Any]) {
        wzb_drawList(with: rect, columns: columns, rows: rows, datas: datas, colorInfo: nil)
    }

    @objc(wzb_drawListWithRect:columns:rows:datas:columnsInfo:)
    func wzb_drawList(
        with rect: CGRect,
        columns: Int,
        rows: Int,
        datas: [Any],
        columnsInfo: [Int: Int]?
    ) {
        wzb_drawList(with: rect, columns: columns, rows: rows, datas: datas, colorInfo: nil, columnsInfo: columnsInfo)
    }

    @objc(wzb_drawListWithRect:columns:rows:datas:colorInfo:)
    func wzb_drawList(
        with rect: CGRect,
        columns: Int,
        rows: Int,
        datas: [Any],
        colorInfo: [Int: UIColor]?
    ) {
        wzb_drawList(with: rect, columns: columns, rows: rows, datas: datas, colorInfo: colorInfo, columnsInfo: nil)
    }

    @objc(wzb_drawListWithRect:columns:rows:datas:colorInfo:columnsInfo:)
    func wzb_drawList(
        with rect: CGRect,
        columns: Int,
        rows: Int,
        datas: [Any],
        colorInfo: [Int: UIColor]?,
        columnsInfo: [Int: Int]?
    ) {
        wzb_drawList(
            with: rect,
            columns: columns,
            rows: rows,
            datas: datas,
            colorInfo: colorInfo,
            columnsInfo: columnsInfo,
            backgroundColorInfo: nil
        )
    }

    @objc(wzb_drawListWithRect:columns:rows:datas:colorInfo:columnsInfo:backgroundColorInfo:)
    func wzb_drawList(
        with rect: CGRect,
        columns: Int,
        rows: Int,
        datas: [Any],
        colorInfo: [Int: UIColor]?,
        columnsInfo: [Int: Int]?,
        backgroundColorInfo: [Int: UIColor]?
    ) {
        wzb_drawList(
            with: rect,
            defaultColumns: columns,
            rows: rows,
            datas: datas,
            colorInfo: colorInfo,
            columnsInfo: columnsInfo,
            columnWidthInfo: nil,
            rowHeights: nil,
            backgroundColorInfo: backgroundColorInfo
        )
    }

    func wzb_drawList(
        with rect: CGRect,
        defaultColumns: Int,
        rowHeights: [CGFloat],
        datas: [Any],
        colorInfo: [Int: UIColor]? = nil,
        columnsInfo: [Int: Int]? = nil,
        columnWidthInfo: [Int: [CGFloat]]? = nil,
        backgroundColorInfo: [Int: UIColor]? = nil
    ) {
        wzb_drawList(
            with: rect,
            defaultColumns: defaultColumns,
            rows: rowHeights.count,
            datas: datas,
            colorInfo: colorInfo,
            columnsInfo: columnsInfo,
            columnWidthInfo: columnWidthInfo,
            rowHeights: rowHeights,
            backgroundColorInfo: backgroundColorInfo
        )
    }

    @objc(getLabelWithIndex:)
    func getLabel(withIndex index: Int) -> UILabel? {
        guard index >= 0, index < wzb_gridStorage.labels.count else { return nil }
        return wzb_gridStorage.labels[index]
    }

    @objc(wzb_drawLineWithFrame:lineType:color:lineWidth:)
    func wzb_drawLine(
        with frame: CGRect,
        lineType: WZBLineType,
        color: UIColor,
        lineWidth: CGFloat
    ) {
        guard frame.width > 0 || frame.height > 0 else { return }

        let path = UIBezierPath()
        path.lineWidth = lineWidth
        path.move(to: .zero)
        switch lineType {
        case .horizontal:
            path.addLine(to: CGPoint(x: frame.width, y: 0))
        case .vertical:
            path.addLine(to: CGPoint(x: 0, y: frame.height))
        }

        let lineLayer = CAShapeLayer()
        lineLayer.frame = frame
        lineLayer.path = path.cgPath
        lineLayer.lineWidth = lineWidth
        lineLayer.strokeColor = color.cgColor
        lineLayer.fillColor = UIColor.clear.cgColor
        wzb_storeLineLayer(lineLayer)
    }

    func wzb_removeGrid() {
        wzb_clearGridArtifacts()
    }
}

public extension UIScrollView {
    @discardableResult
    func wzb_drawScrollableList(
        origin: CGPoint = .zero,
        cellSize: CGSize,
        columns: Int,
        rows: Int,
        datas: [Any],
        colorInfo: [Int: UIColor]? = nil,
        columnsInfo: [Int: Int]? = nil,
        backgroundColorInfo: [Int: UIColor]? = nil
    ) -> UIView {
        let storage: WZBScrollableGridStorage
        if let existing = objc_getAssociatedObject(self, &WZBAssociatedKeys.scrollableGridStorage) as? WZBScrollableGridStorage {
            storage = existing
        } else {
            storage = WZBScrollableGridStorage()
            objc_setAssociatedObject(
                self,
                &WZBAssociatedKeys.scrollableGridStorage,
                storage,
                .OBJC_ASSOCIATION_RETAIN_NONATOMIC
            )
        }

        storage.containerView?.removeFromSuperview()

        let resolvedColumnsInfo = columnsInfo ?? [:]
        let contentWidth = (0 ..< rows)
            .map { max(resolvedColumnsInfo[$0] ?? columns, 1) }
            .map { CGFloat($0) * cellSize.width }
            .max() ?? cellSize.width
        let contentHeight = CGFloat(rows) * cellSize.height

        let containerView = UIView(frame: CGRect(
            x: origin.x,
            y: origin.y,
            width: contentWidth,
            height: contentHeight
        ))
        addSubview(containerView)

        containerView.wzb_drawList(
            with: containerView.bounds,
            columns: columns,
            rows: rows,
            datas: datas,
            colorInfo: colorInfo,
            columnsInfo: columnsInfo,
            backgroundColorInfo: backgroundColorInfo
        )

        storage.containerView = containerView
        contentSize = CGSize(
            width: max(bounds.width, origin.x + contentWidth),
            height: max(bounds.height, origin.y + contentHeight)
        )
        return containerView
    }
}

private extension UIView {
    func wzb_drawGridBorder(for rect: CGRect) {
        let x = rect.minX
        let y = rect.minY
        let width = rect.width
        let height = rect.height
        let lineWidth = wzb_hairlineWidth

        wzb_drawLine(
            with: CGRect(x: x, y: y, width: width, height: lineWidth),
            lineType: .horizontal,
            color: wzbDefaultSeparatorColor(),
            lineWidth: lineWidth
        )
        wzb_drawLine(
            with: CGRect(x: x + width - lineWidth, y: y, width: lineWidth, height: height),
            lineType: .vertical,
            color: wzbDefaultSeparatorColor(),
            lineWidth: lineWidth
        )
        wzb_drawLine(
            with: CGRect(x: x, y: y + height - lineWidth, width: width, height: lineWidth),
            lineType: .horizontal,
            color: wzbDefaultSeparatorColor(),
            lineWidth: lineWidth
        )
        wzb_drawLine(
            with: CGRect(x: x, y: y, width: lineWidth, height: height),
            lineType: .vertical,
            color: wzbDefaultSeparatorColor(),
            lineWidth: lineWidth
        )
    }
}
